cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1066,20 +1113,95 @@ class ObjectAppendStream final : public io::OutputStream {
       // flush. This also avoids some unhandled errors when flushing in the destructor.
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    std::unique_lock<std::mutex> lock(upload_state_->mutex);
+    return CommitBlockList(block_blob_client_, upload_state_->block_ids,
+                           commit_block_list_options_);
```

Comment:
If we restructure the lock as you've suggested, we get a deadlock because the future can actually never be marked as finished by a succeeding background write (because in `HandleUploadOutcome`, where we try to mark the future as finished, we also need to acquire the lock).
I think the race you pointed out is only an issue if there is a parallel `Write()` and `Flush()` to the stream. That is the only way how `pending_blocks_completed` could change the state during `Flush()`. To my knowledge, arrow's output streams are not supposed to be thread-safe (e.g. the S3 output stream isn't), so it should not be a problem in practice.

Reply:
This is very important information that should be in the source code. Multi-threaded code should have comments containing arguments for its correctness.

Reply:
I added a comment to indicate that it is not thread-safe. [2ffbcb3](https://github.com/apache/arrow/pull/43096/commits/2ffbcb3e644290b3a66270392c50593d59d5bab8)